src/light_data.rs

```diff
@@ -0,0 +1,307 @@
+            Self::Cell(xs) if xs.len() <= 64 => {
+                for x in xs {
+                    res.extend(x.ser())
+                }
+            }
+            Self::Cell(xs) => {
+                res.extend(LightData::to_trimmed_le_bytes(xs.len()));
+                for x in xs {
+                    res.extend(x.ser())
+                }
+            }
+        };
+        res
+    }
+
+    pub fn de(i: &[u8]) -> Result<Self, ()> {
+        match Self::de_aux(i).finish() {
+            Ok((_, x)) => Ok(x),
+            Err(_) => Err(()),
+        }
+    }
+
+    pub fn de_aux(i: &[u8]) -> IResult<&[u8], Self> {
+        let (i, tag) = take(1u8)(i)?;
+        let tag = tag[0];
+        let size = tag & 0b11_1111;
+        if Self::tag_is_atom(tag) {
+            let (i, size) = if Self::tag_is_small(tag) && size == 0 {
+                Ok((i, 64))
+            } else if Self::tag_is_small(tag) {
+                Ok((i, size as usize))
+            } else {
+                let (i, size) = take(size)(i)?;
+                let size = size.iter().fold(0, |acc, &x| (acc * 256) + x as usize);
+                Ok((i, size as usize))
+            }?;
+            let (i, xs) = take(size)(i)?;
+            Ok((i, LightData::Atom(xs.to_vec())))
+        } else {
+            let (i, size) = if Self::tag_is_small(tag) && size == 0 {
+                Ok((i, 64))
+            } else if Self::tag_is_small(tag) {
+                Ok((i, size as usize))
+            } else {
+                let (i, size) = take(size)(i)?;
+                let size = size.iter().fold(0, |acc, &x| (acc * 256) + x as usize);
+                Ok((i, size as usize))
+            }?;
+            let (i, xs) = count(LightData::de_aux, size)(i)?;
+            Ok((i, LightData::Cell(xs.to_vec())))
+        }
+    }
+}
+
+pub trait Encodable {
+    fn ser(&self) -> LightData;
+    fn de(ld: &LightData) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+impl<A: Encodable + Sized> Encodable for Option<A> {
+    fn ser(&self) -> LightData {
+        match self {
+            None => LightData::Cell(vec![]),
+            Some(a) => LightData::Cell(vec![a.ser()]),
+        }
+    }
+
+    fn de(ld: &LightData) -> Result<Self, String> {
+        match ld {
+            LightData::Cell(xs) => match xs.as_slice() {
+                [] => Ok(Option::None),
+                [a] => Ok(Option::Some(A::de(a)?)),
+                _ => Err(format!("expected Option")),
+            },
+            _ => Err(format!("expected Option")),
+        }
+    }
+}
+
+impl<A: Encodable + Sized> Encodable for Vec<A> {
+    fn ser(&self) -> LightData {
+        LightData::Cell(self.iter().map(|x| x.ser()).collect())
+    }
+
+    fn de(ld: &LightData) -> Result<Self, String> {
+        match ld {
+            LightData::Cell(xs) => {
+                let mut res = vec![];
+                for x in xs {
+                    res.push(A::de(x)?)
+                }
+                Ok(res)
```

Comment:
This is an instance of [iterating over results](https://rustwiki.org/en/rust-by-example/error/iter_result.html) and could be expressed more simply.